### PR TITLE
logs now report as pvgroups are added to pvl logs (fixes #4914)

### LINKED ISSUE #4914: Apps converted to a callable no longer log in real time
**ISIS version(s) affected**: varies by app. Conversion started at 3.7, so 3.7+

**Description**  

Apps that have been converted to a callable function no longer log directly to the application, instead they log to a PVL object and then upon completion log its whole contents to the application.

**How to reproduce**  

This is most notable with jigsaw. If you run jigsaw in ISIS 4.4 or later it won't log until it reaches the very end of the process.

**Possible Solution**  

One possible solution is to create a logger object that logs things, if possible, as we add to it. This would be similar to the current Application::log function, but it would be an object passed into function calls similar to our process objects.

### PR DESCRIPTION
**Pull request by Kelvinrr**
## Description

Pvl logs were being reported at the end of the programs execution which is a problem for apps that run for a long time. This changes the apps to report as groups are added.

This changes Application::Log to work with or without an iApp instance. Plus, a new function `addLogGroup` that adds the PVL group as usual but also calls Application::Log on the object. 

Tried to catch all the converted apps using a recursive grep for `Pvl appLog`, `Pvl results`, `Pvl log` in `main.cpp`s. I think I got all or at least a vast majority of them. 

`Application::Log` calls were removed in `main.cpp`s of converted tests as they're no longer needed. Left `Application::GuiLog` calls were left alone. Getsn was also left alone since it had very specific logging code in both the main and callable. 

## Related Issue

fixes #4914

## Motivation and Context

Long running apps like jigsaw were not getting realtime updates. 

## How Has This Been Tested?

Jenkins tests, if nothing broke no new test failures should appear. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

**Comment by Kelvinrr** [comment]
I'll update the test writing docs on the wiki after this is merged

File: isis/src/base/apps/stretch/stretch_app.h [hunk 1]
```diff
@@ -4,6 +4,7 @@
 #include "PvlGroup.h"
 #include "PvlKeyword.h"
 #include "UserInterface.h"
+#include "PvlLog.h"
```
Comment (AustinSanders) [inline isis/src/base/apps/stretch/stretch_app.h]: Where does PvlLog.h come from?  I can't find it anywhere

**Review by AustinSanders — COMMENTED** [review]
Overall I think it looks good.  I can't find PvlLog.h, so I'm not sure what it contains.  I hard a hard time understanding which files needed PvlLog.h imports as opposed to simple Pvl imports.

File: isis/src/base/apps/stretch/stretch_app.h (hunk 1 above)
Comment (AustinSanders) [reply isis/src/base/apps/stretch/stretch_app.h]: Not all of the updated apps import PvlLog.h -- when is it necessary / unnecessary?

Comment (Kelvinrr) [reply isis/src/base/apps/stretch/stretch_app.h]: left over from a previous approach I ditched. Should be removed.

File: isis/src/base/objs/Application/Application.h [hunk 2]
```diff
@@ -31,6 +31,7 @@ namespace Isis {
   class Gui;
   class IException;
   class Progress;
+  class PvlObject;
```
Comment (jessemapel) [inline isis/src/base/objs/Application/Application.h]: Is this needed?

File: isis/src/base/objs/PvlObject/PvlObject.cpp [hunk 3]
```diff
@@ -12,6 +12,7 @@ find files of those names at the top level of this repository. **/
 #include "IString.h"
 #include "Message.h"
 #include "PvlFormat.h"
+#include "Application.h"
```
Comment (jessemapel) [inline isis/src/base/objs/PvlObject/PvlObject.cpp:15]: I'm not super happy with PvlObject being dependent on Application, but I think this is the cleanest solution.

**Review by jessemapel — COMMENTED** [review]
Also needs a CHANGELOG entry

File: isis/src/base/objs/PvlObject/PvlObject.cpp (hunk 3 above)
Comment (Kelvinrr) [reply isis/src/base/objs/PvlObject/PvlObject.cpp:15]: Yeah, I'm not super keen on it either.

File: isis/src/base/objs/Application/Application.h (hunk 2 above)
Comment (Kelvinrr) [reply isis/src/base/objs/Application/Application.h]: left over from the class approach, removed it

**Comment by Kelvinrr** [comment]
results from local tests: 

```
The following tests FAILED:
        543 - FitsToJson.FitsConversion (Subprocess aborted)
        623 - TempTestingFiles.UnitTestImagePolygonCross (Failed)
        853 - LronaccalDefault.FunctionalTestsLronaccal (Failed)
        854 - LronaccalNear.FunctionalTestsLronaccal (Failed)
        855 - LronaccalPair.FunctionalTestsLronaccal (Failed)
```

They're all weird "cant find data file" type errors, most likely not related to these changes at all.

File: isis/src/base/objs/PvlObject/PvlObject.cpp (hunk 3 above)
Comment (acpaquette) [reply isis/src/base/objs/PvlObject/PvlObject.cpp:15]: This is spooky for integrating the ISIS Blob work next year. Likely requiring that Application and everything associated with it get pulled into that submodule within ISIS

**Comment by Kelvinrr** [comment]
Gonna rerun all tests after that ctest PR was merged and double check everything is fine.

**Comment by Kelvinrr** [comment]
this time I got: 

```
The following tests FAILED:
	520 - FitsToJson.FitsConversion (Subprocess aborted)
	600 - TempTestingFiles.UnitTestImagePolygonCross (Failed)
	899 - TestKernelDb.TestKernelsSmithOffset (Failed)
```

520 - some weird QT assertion error within the QT lib, most likely something weird with the version of QT my env picked up 
600 - that tests seems to have a bad path for it's data? It doesn't like that my build dir is outside the repo it seems. 
899 - weird CK error, seems like the same as 600 

Either way, these don't seem to have anything to do with my changes and more to do with my build dir and env on prog28.

File: isis/src/base/objs/PvlObject/PvlObject.cpp (hunk 3 above)
Comment (jessemapel) [reply isis/src/base/objs/PvlObject/PvlObject.cpp:15]: Based on some discussion this morning. It's not ideal, but this solution is still good and the best one we have right now. We will need to come back and re-address some of the architecture next year so we're just adding some tech debt for us to deal with next year.

**Review by jessemapel — APPROVED** [review]

**Pull request merged**